Feat: show connected wallet info

DimaKush opened the pull request:
# Description

Added builder status indicators (batch membership and check-in status) to the wallet connection dropdown.

## Changes
- Created useBuilderStatus hook to fetch status from BatchRegistry contract
- Added status icons to AddressInfoDropdown component
- Shows green icons for allowlisted and checked-in addresses

## Testing
1. Connect wallet
2. Verify batch member icon (group) appears if address is allowlisted
3. Verify check-in icon appears if address has checked in

## Screenshot
![Screenshot 2025-01-09 at 19-32-55 Debug Contracts Scaffold-ETH 2](https://github.com/user-attachments/assets/833f9ee7-ddde-4ef8-a5c6-397c835af57b)

## Related Issues

Closes #7

This PR includes:
- The new useBuilderStatus hook
- Modified AddressInfoDropdown component
- Proper imports and status indicators

The changes are focused and minimal, adding the requested functionality without unnecessary modifications.

ENS/address: 0x845D62b5405aAAe6b95CF3fbf33968A1E71a9cAD

derrekcoleman commented:
Thank you for the PR, @DimaKush! I appreciate your detailed description and screenshot, very professional.

Let's take a step back and consider the goal of the issue. The idea is to give the user visual feedback for whether they are (1) on the `allowList` and (2) `checkedIn` or not. Your current implementation shows these icons _inside of_ the wallet component, making it impossible (or at least not obvious) to hover these icons and _discover_ what they are trying to say. Maybe moving them next to (but outside of) the wallet component would make this UX more discoverable.

Thinking about state for a moment, there are several possible paths:
- No wallet connected (we want user to connect)
- Connected, not on `allowList` (we want to let them know)
- Connected, on `allowList`, not `checkedIn` (we want them to `checkIn`)
- Connected, on `allowList`, `checkedIn` (we want them to feel successful)

How might you represent each of these states in a way that helps the user discover the next step they need to take? Right now you show green icons for `true` states, but what about "failing gracefully" and guiding the user for `false` states?

FrankiePower commented:
@DimaKush i think you did a good job. but i have an idea for something. if you want you can let me do it or make improvement to yours or do it instead

![Screenshot from 2025-01-10 09-34-58](https://github.com/user-attachments/assets/cd40c8e6-0e86-4c79-a534-3109b72d8bcf)

@derrekcoleman if @DimaKush is okay with it, i can do this. or we could just go which whatever he has. this is just the idea  i had for it and i thought to share

DimaKush commented:
@FrankiePower Yep looks good, but on mobile we don't have a lot of space. So I moved this to main page.

FrankiePower commented:
okay @DimaKush . also if you don't mind i'd like to see a preview (screenshot) of what it looks like

DimaKush commented:
![image](https://github.com/user-attachments/assets/90a8ad7c-9cec-44e3-8a72-90078158e6d0)
![image](https://github.com/user-attachments/assets/f13aa65f-eadf-4acb-9123-651f47197cc3)
![image](https://github.com/user-attachments/assets/d3aab77f-7c87-42ad-a4bb-c37dbf77f089)

All-Khwarizmi commented:
I'm not working on this but reading through your work I thought we should mix it together.

Since I would prefer the current user status information to be displayed in the header, it would be nice to have a more concise unique badge, instead of two, that is responsible for displaying the user status. 

https://github.com/user-attachments/assets/a52a13bb-d698-4649-81ad-91b4b2c77420

What do you think?

FrankiePower commented:
i think i also like what @All-Khwarizmi is proposing. we can find a way to implement the responsiveness for other screens. but then you'd also have to find a way to display what @DimaKush is trying to display.

derrekcoleman commented:
I love this conversation! This design "problem" appears all of the time - we want the information to be obvious/available (e.g. in the header of all pages), but need it to be concise/unobtrusive (e.g. make sense on mobile).

Personally, I like the pattern of using attention-grabbing icons (imagine ❗ for error states) with on-hover/on-click tooltips that the user can easily discover to learn more in 1-2 sentences.

DimaKush commented:
## Move Status Component to Header

Moved batch status from homepage to header for better UX. Users can now see their allowlist/check-in status from any page.

### Before
- Status only visible on homepage
- Had to navigate back to check status
- Easy to miss important state changes

### After
- Always visible in header
- Quick access to GitHub/batch links
- Immediate feedback on status changes
- Works well on mobile/desktop

### Changes
- Moved `UserStateInBatch` component into Header

### Screenshots
![Screenshot from 2025-01-12 11-50-05](https://github.com/user-attachments/assets/bd77578b-1134-49c7-bb0a-92ab52cf8b01)
![Screenshot from 2025-01-12 11-50-23](https://github.com/user-attachments/assets/e6d0692f-1596-4a12-a808-2c9f241adf40)
![Screenshot from 2025-01-12 11-51-46](https://github.com/user-attachments/assets/fa27c8bd-a729-4272-af4f-9cb195ec6292)

derrekcoleman reviewed (changes requested):
Looking really nice! Thanks for incorporating all of our feedback, @DimaKush. I think the URLs are a really nice touch 👌 

Could you please use `git rebase` to `drop` the first commit related to your personal page? Once it's gone from the commit history, you can `git push --force` to update this PR as if it was never there. This is your chance to practice git-fu 🥷

derrekcoleman commented on packages/nextjs/components/UserStateInBatch.tsx:
```diff
@@ -0,0 +1,81 @@
+              <div>
+                You need to check in to start building. Follow instructions in{" "}
+                <a
+                  href="https://github.com/BuidlGuidl/batch12.buidlguidl.com/issues/12"
+                  target="_blank"
+                  className="underline"
+                >
+                  this GitHub issue
+                </a>
+              </div>,
+            )
+          }
+        >
+          <CheckCircleIcon className="h-6 w-6 text-warning cursor-pointer hover:text-warning/70" />
+        </div>
+      )}
+
+      {isCheckedIn && (
+        <div
+          className="tooltip tooltip-bottom"
+          data-tip="You are checked in and ready to build!"
+          onClick={() =>
+            notification.success(
+              <div>
+                You are checked in and ready to build! Check{" "}
+                <a
+                  href="https://github.com/BuidlGuidl/batch12.buidlguidl.com/issues"
+                  target="_blank"
+                  className="underline"
+                >
+                  GitHub issues
+                </a>{" "}
+                for next steps.
+              </div>,
+            )
+          }
+        >
+          <CheckCircleIcon className="h-6 w-6 text-success cursor-pointer hover:text-success/70" />
+        </div>
+      )}
```
I see some repetition in this logic, which is a good opportunity to refactor the data away from the display logic. 

It seems like there are two parts that could be refactored out into reusable components: 
1. The logic for checking state and displaying an icon
2. The logic for handling the notification you pop up `onClick`

Please refactor this section to be more [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) 🙏

DimaKush replied:
refactored in [commit](https://github.com/BuidlGuidl/batch12.buidlguidl.com/pull/22/commits/b9e3dcddf7973f842779aabd332fe2ffe76bda31)

derrekcoleman commented on packages/nextjs/components/UserStateInBatch.tsx:
```diff
@@ -0,0 +1,81 @@
+  const { address } = useAccount();
+  const { isAllowListed, isCheckedIn } = useBuilderStatus(address);
+
+  if (!address) return null;
```
`null` is a bit scary, since it doesn't [fail gracefully](https://en.wikipedia.org/wiki/Graceful_exit) and help the user understand what's happening in case of an error. Since this component returns JSX, what would be reasonable to render when `!address`?

DimaKush replied:
Resolved in [commit](https://github.com/BuidlGuidl/batch12.buidlguidl.com/pull/22/commits/b9e3dcddf7973f842779aabd332fe2ffe76bda31#diff-62c77dddc478e699f89c0c8b265d3ed3dc21dcc33919d7706aed0ac02152aaac)

derrekcoleman reviewed (changes requested):
Looking great! I love the way you've factored it.

Now I'm really pushing you to polish the core code and get it production-ready (i.e. easier to maintain later) before we merge 🏋️

DimaKush commented:
refactor: improve wallet disconnected state UX and reduce code duplication

- Replace null return with disconnected state UI
- Add wallet icon and config for disconnected state
- Show helpful message when wallet isn't connected
- Extract builder state logic into custom hook useBuilderState
- Add warning type to notification system
- Reduce repetition in state/notification management
- Improve component reusability by decoupling state and display logic
- Move state configs to single file for better maintainability
- Centralize builder state type definitions

![image](https://github.com/user-attachments/assets/2ff5a32b-c296-438a-9bc2-572f51ff929e)

derrekcoleman reviewed (changes requested):
Thanks for the updates, @DimaKush! It's looking very solid.

Found a little UI bug where there isn't any spacing when wallet connected and is on "wrong network" (not OP Mainnet)
<img width="342" alt="Screenshot 2025-01-16 at 12 33 33 PM" src="https://github.com/user-attachments/assets/af6e7da6-f674-4035-a0c9-5bab21b61fae" />

Please remember to respond to and resolve previous review comments as you address them. It makes it much easier for the review (or other people taking a look at the PR) to come back and see what has changed since the last review.

DimaKush commented:
Thank you for feedback, @derrekcoleman !


> Thanks for the updates, @DimaKush! It's looking very solid.
> 
> Found a little UI bug where there isn't any spacing when wallet connected and is on "wrong network" (not OP Mainnet) <img alt="Screenshot 2025-01-16 at 12 33 33 PM" width="342" src="https://private-user-images.githubusercontent.com/80121818/404015189-af6e7da6-f674-4035-a0c9-5bab21b61fae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczNDMwMzAsIm5iZiI6MTczNzM0MjczMCwicGF0aCI6Ii84MDEyMTgxOC80MDQwMTUxODktYWY2ZTdkYTYtZjY3NC00MDM1LWEwYzktNWJhYjIxYjYxZmFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIwVDAzMTIxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE1MDM0NmU3MGQxMzdjYzk3MzRjNWE0YzMzYWY0NjkzNDU3YmM3YzFhYTYwNmIwNzM0ZDRmZjIyZDYxMDA0MmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xEcNWtgsDZvaS8nfmM7Si0tDKDOK14ZyFx348otB0bw">
> 
> Please remember to respond to and resolve previous review comments as you address them. It makes it much easier for the review (or other people taking a look at the PR) to come back and see what has changed since the last review.

Summarize of all changes in this PR:

# Display User State In Batch Feature

This PR adds comprehensive user state display functionality with improved UX and code organization.

## Purpose

This feature helps users understand their current state in the builder system and guides them through the necessary steps to participate.

## Files Changed & Their Purpose

### 🎯 Core Logic
#### `hooks/useBuilderState.ts`
- **Purpose**: Central source of truth for builder state management
- Handles all state checks (wallet, network, allowlist, check-in)
- Provides type-safe state configs and notifications
- Single responsibility: determine and provide user's current state

### 🎨 UI Components
#### `components/builderState/BuilderStateIcon.tsx`
- **Purpose**: Visual indicator for user's current state
- Shows different icons based on state (wallet, check mark, group)
- Provides consistent visual feedback through colors and animations
- Handles hover tooltips for quick status understanding

#### `components/builderState/BuilderNotification.tsx`
- **Purpose**: Interactive feedback system
- Shows actionable notifications with relevant links
- Guides users through required steps
- Provides clear feedback for state changes

#### `components/BuilderStateInBatch.tsx`
- **Purpose**: Main container component
- Orchestrates state and UI components
- Handles user interactions
- Maintains consistent layout across states

## Technical Details
- DRY code with shared styles and configs
- Type-safe state management
- Proper error handling
- Consistent styling

derrekcoleman commented on packages/nextjs/components/builderState/BuilderStateIcon.tsx:
```diff
@@ -2,17 +2,22 @@ import { WalletIcon } from "@heroicons/react/24/outline";
 };
 
 export const BuilderStateIcon = ({ type, onClick, tooltip }: BuilderStateIconProps) => {
+  const baseStyles = "h-6 w-6 cursor-pointer";
+  const warningStyles = `${baseStyles} text-warning animate-pulse`;
+  const walletStyles = `${warningStyles} mr-2`;
```
Tailwind purists will say not to do it this way, but I think this kind of "inheritance" pattern is easy to read, understand, and maintain in this case!

derrekcoleman commented on packages/nextjs/components/builderState/BuilderStateIcon.tsx:
```diff
@@ -2,17 +2,22 @@ import { WalletIcon } from "@heroicons/react/24/outline";
-    notCheckedIn: <CheckCircleIcon className="h-6 w-6 text-warning animate-pulse cursor-pointer" />,
-    checkedIn: <CheckCircleIcon className="h-6 w-6 text-success cursor-pointer" />,
-    disconnected: <WalletIcon className="h-6 w-6 text-warning animate-pulse cursor-pointer mr-2" />,
+    notAllowlisted: <UserGroupIcon className={warningStyles} />,
+    notCheckedIn: <CheckCircleIcon className={warningStyles} />,
+    checkedIn: <CheckCircleIcon className={`${baseStyles} text-success`} />,
+    disconnected: <WalletIcon className={walletStyles} />,
+    wrongNetwork: <WalletIcon className={walletStyles} />,
```
![image](https://github.com/user-attachments/assets/8367928d-c1f1-4359-b728-0b8d0fd222cb)

derrekcoleman reviewed (approved):
I'm proud of this PR, DimaKush - and I hope you are too. Great code organization across and within files and good job covering lots of edge-cases in a consistent way.
